src/main/java/de/dafuqs/spectrum/blocks/fluid/RemainderlessItemFluidStorage.java

```diff
@@ -0,0 +1,82 @@
+package de.dafuqs.spectrum.blocks.fluid;
+
+import net.fabricmc.fabric.api.transfer.v1.context.ContainerItemContext;
+import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant;
+import net.fabricmc.fabric.api.transfer.v1.item.ItemVariant;
+import net.fabricmc.fabric.api.transfer.v1.storage.StoragePreconditions;
+import net.fabricmc.fabric.api.transfer.v1.storage.base.ExtractionOnlyStorage;
+import net.fabricmc.fabric.api.transfer.v1.storage.base.SingleSlotStorage;
+import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext;
+import net.minecraft.item.Item;
+
+
+@SuppressWarnings("UnstableApiUsage")
+public class RemainderlessItemFluidStorage implements ExtractionOnlyStorage<FluidVariant>, SingleSlotStorage<FluidVariant> {
+
+    private final ContainerItemContext context;
+    private final Item fullItem;
+    private final FluidVariant containedFluid;
+    private final long containedAmount;
+
+
+    public RemainderlessItemFluidStorage(ContainerItemContext context, FluidVariant containedFluid, long containedAmount) {
+        StoragePreconditions.notBlankNotNegative(containedFluid, containedAmount);
+
+        this.context = context;
+        this.fullItem = context.getItemVariant().getItem();
+        this.containedFluid = containedFluid;
+        this.containedAmount = containedAmount;
+    }
+
+    @Override
+    public long extract(FluidVariant resource, long maxAmount, TransactionContext transaction) {
+        // If the context's item is not fullItem anymore, can't extract!
+        if (!context.getItemVariant().isOf(fullItem)) return 0;
+
+        // Make sure that the fluid and the amount match.
+        if (resource.equals(containedFluid) && maxAmount >= containedAmount) {
```

Comment:
This isn't really solvable as it is now. It can only be worked around by e.g. extracting the fluid first, and then using a secondary storage containing the resultant fluid for further processing.

Reply:
For now this should be fine as it only happens with Mermaids Gems. This type of behavior would apply when using say, a Lava Bucket on a partially filled tank or cauldron.